File: backend/app/Models/SavingsGoal.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

class SavingsGoal extends Model
{
    use HasFactory;

    protected $fillable = ['user_id', 'name', 'target_amount', 'current_amount', 'deadline', 'is_completed'];

    protected $casts = [
        'deadline'     => 'date',
        'is_completed' => 'boolean',
        'target_amount'  => 'decimal:2',
        'current_amount' => 'decimal:2',
    ];

    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }
}

File: android/app/src/main/kotlin/com/duit/app/domain/model/SavingsGoal.kt
package com.duit.app.domain.model

data class SavingsGoal(
    val id: Int,
    val name: String,
    val targetAmount: Double,
    val currentAmount: Double,
    val deadline: String?,
    val isCompleted: Boolean
) {
    val progress: Float get() = if (targetAmount <= 0) 0f else (currentAmount / targetAmount).toFloat().coerceIn(0f, 1f)
}
